crypto/cms/cms_env.c

```diff
@@ -169,45 +235,34 @@ CMS_ContentInfo *CMS_EnvelopedData_create(const EVP_CIPHER *cipher)
     return CMS_EnvelopedData_create_with_libctx(cipher, NULL, NULL);
 }
 
-int cms_EnvelopedData_final(CMS_ContentInfo *cms, BIO *chain)
+CMS_ContentInfo *
+CMS_AuthEnvelopedData_create_with_libctx(const EVP_CIPHER *cipher,
+                                         OPENSSL_CTX *libctx,
+                                         const char *propq)
 {
-    CMS_EnvelopedData *env = NULL;
-    EVP_CIPHER_CTX *ctx = NULL;
-    BIO *mbio = BIO_find_type(chain, BIO_TYPE_CIPHER);
-
-    env = cms_get0_enveloped(cms);
-    if (env == NULL)
-        return 0;
-
-    if (mbio == NULL) {
-        CMSerr(CMS_F_CMS_ENVELOPEDDATA_FINAL, CMS_R_CONTENT_NOT_FOUND);
-        return 0;
-    }
-
-    BIO_get_cipher_ctx(mbio, &ctx);
+    CMS_ContentInfo *cms;
+    CMS_AuthEnvelopedData *aenv;
 
-    /*
-     * If the selected cipher supports unprotected attributes,
-     * deal with it using special ctrl function
-     */
-    if (EVP_CIPHER_flags(EVP_CIPHER_CTX_cipher(ctx)) & EVP_CIPH_FLAG_CIPHER_WITH_MAC) {
-        if (cms->d.envelopedData->unprotectedAttrs == NULL)
-            cms->d.envelopedData->unprotectedAttrs = sk_X509_ATTRIBUTE_new_null();
+    cms = CMS_ContentInfo_new_with_libctx(libctx, propq);
+    if (cms == NULL)
+        goto merr;
+    aenv = cms_auth_enveloped_data_init(cms);
+    if (aenv == NULL)
+        goto merr;
+    if (!cms_EncryptedContent_init(aenv->authEncryptedContentInfo,
+                                   cipher, NULL, 0, cms_get0_cmsctx(cms)))
+        goto merr;
+    return cms;
+ merr:
+    CMS_ContentInfo_free(cms);
+    CMSerr(0, ERR_R_MALLOC_FAILURE);
+    return NULL;
+}
 
-        if (cms->d.envelopedData->unprotectedAttrs == NULL) {
-            CMSerr(CMS_F_CMS_ENVELOPEDDATA_FINAL, ERR_R_MALLOC_FAILURE);
-            return 0;
-        }
 
-        if (EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_PROCESS_UNPROTECTED,
-                                1, env->unprotectedAttrs) <= 0) {
-            CMSerr(CMS_F_CMS_ENVELOPEDDATA_FINAL, CMS_R_CTRL_FAILURE);
-            return 0;
-        }
-    }
-
-    cms_env_set_version(cms->d.envelopedData);
-    return 1;
+CMS_ContentInfo *CMS_AuthEnvelopedData_create(const EVP_CIPHER *cipher)
```

Comment:
If this is a completely new API.. it probably doesn't need this variant at all..
The only reason the other ones are there is because they already existed.

Reply:
But this allows using `CMS_AuthEnvelopedData_create` without using the context which is much nicer for users IMO. This API is still useful and it's consistent with `CMS_EnvelopedData_create`.

Reply:
ok..